File: 백준/Gold/1918. 후위 표기식/후위 표기식.py
stack = []
priority = {"+": 1, "-": 1, "*": 2, "/": 2, "(":0}

for c in input():
    if c.isalpha():
        print(c, end="")
        continue

    if c == "(":
        stack.append(c)
    elif c == ")":
        while stack and stack[-1] != "(":
            print(stack.pop(), end="")
        stack.pop() # (는 출력 대상 아님
    else:
        while stack and priority[stack[-1]] >= priority[c]:
            print(stack.pop(), end="")
        stack.append(c)

while stack:
    print(stack.pop(), end="")

File: 백준/Gold/1918. 후위 표기식/후위 표기식.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String input = sc.nextLine();

        Stack<Character> stack = new Stack<>();
        Map<Character, Integer> priority = new HashMap<>();
        priority.put('+', 1);
        priority.put('-', 1);
        priority.put('*', 2);
        priority.put('/', 2);
        priority.put('(', 0);

        StringBuilder result = new StringBuilder();

        for (char c : input.toCharArray()) {
            if (Character.isLetter(c)) { 
                result.append(c);
                continue;
            }

            if (c == ')') {
                while (!stack.isEmpty() && stack.peek() != '(') {
                    result.append(stack.pop());
                }
                stack.pop();
                continue;
            }

            if (c == '(') {
                stack.push(c);
                continue;
            }

            while (!stack.isEmpty() && priority.get(c) <= priority.get(stack.peek())) {
                result.append(stack.pop());
            }
            stack.push(c);
        }

        while (!stack.isEmpty()) {
            result.append(stack.pop());
        }

        System.out.println(result.toString());
    }
}
